<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="utf-8">
    <meta name="HandheldFriendly" content="True">
    <meta name="MobileOptimized" content="320">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="referrer" content="no-referrer">
    <meta name="description" content="A website built through Hugo and blogdown.">

    <link rel="stylesheet" href="https://code.jquery.com/ui/1.14.1/themes/base/jquery-ui.css">
    <script src="https://code.jquery.com/jquery-1.12.4.js"></script>
    <script src="https://code.jquery.com/ui/1.12.1/jquery-ui.js"></script>


    
      <link href="https://fonts.googleapis.com/css?family=Open+Sans:400|Old+Standard+TT:400&display=swap" rel="stylesheet" media="print" type="text/css" onload="this.media='all'">
    

    


<link href='//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css' rel='stylesheet' type='text/css' />




    <title>
      
      
         Visualize relations between genes and pathways 
      
    </title>
    <link rel="canonical" href="lab-cn/2020/11/27/visualize-relations-between-genes-and-pathways/">

    <style>
  * {
    border:0;
    font:inherit;
    vertical-align:baseline;
    margin:0;
    padding:0;
    color: black;
    text-decoration-skip: ink;
    overflow-wrap: break-word;
  }

  body {
    font-family: Georgia,serif;
    line-height:160%;
    color:#1d1313;
    max-width:1000px;
    margin:auto;
    min-height: 600px;
    font-size: 17px;
  }

  p {
    margin: 20px 0;
  }

  a img {
    border:none;
  }

  img {
    margin: 10px auto 10px auto;
    max-width: 100%;
    display: block;
  }

  .left-justify { float: left; }
  .right-justify { float:right; }

  pre, code {
    font: 12px Consolas, "Liberation Mono", Menlo, Courier, monospace;
    background-color: #f7f7f7;
  }

  code {
    padding: 4px;
  }

  pre {
    margin-top: 0;
    margin-bottom: 16px;
    word-wrap: normal;
    padding: 16px;
    overflow: auto;
    font-size: 85%;
    line-height: 1.45;
  }

  pre>code {
    padding: 0;
    margin: 0;
    font-size: 100%;
    word-break: normal;
    white-space: pre;
    background: transparent;
    border: 0;
  }

  pre code {
    display: inline;
    padding: 0;
    margin: 0;
    overflow: visible;
    line-height: inherit;
    word-wrap: normal;
    background-color: transparent;
    border: 0;
  }

  pre code::before,
  pre code::after {
    content: normal;
  }

  em,q,em,dfn {
    font-style:italic;
  }

  .sans,html .gist .gist-file .gist-meta {
    font-family:"Open Sans","Myriad Pro",Myriad,sans-serif;
  }

  .mono,pre,code,tt,p code,li code {
    font-family:Menlo,Monaco,"Andale Mono","lucida console","Courier New",monospace;
  }

  strong {
    font-weight:600;
  }

  q:before {
    content:"\201C";
  }

  q:after {
    content:"\201D";
  }

  del,s {
    text-decoration:line-through;
  }

  blockquote {
    font-family:"Old Standard TT",serif;
    text-align:center;
    padding:50px;
  }

  blockquote p {
    display:inline-block;
    font-style:italic;
  }

  blockquote:before,blockquote:after {
    font-family:"Old Standard TT",serif;
    content:'\201C';
    color:#403c3b;
  }

  blockquote:after {
    content:'\201D';
  }

  hr {
    width:40%;
    height: 1px;
    background:#403c3b;
    margin: 25px auto;
  }


  h2, h3 {
    margin-top: 40px;
    margin-bottom: 20px;
  }

  h1 a,h2 a,h3 a {
    text-decoration:none;
  }

  h1,h2 {
    margin-top:28px;
  }

  h1 {
    font-size: 2em;
  }

  h2, h3 {
    font-size: 1.5em;
  }

  #sub-header, .date {
    color:#403c3b;
  }

  #sub-header {
    margin: 0 4px;
    font-style: italic;
  }

  #nav h1 a {
    color:#1d1313;
    line-height:120%;
  }

  .posts_listing a,#nav a {
    text-decoration: none;
  }

  li {
    margin-left: 20px;
  }

  ul li {
    margin-left: 20px;
    padding-top: 2px;
    padding-bottom: 2px;
  }


  #nav ul li:before, .posts_listing li:before {
    content:'';
    margin-right:0;
  }

  #content {
    text-align:left;
    width:100%;
    padding:30px 0 80px;
  }

  #content h1,#content h2 {
    margin-bottom:5px;
  }

  #content .entry-content {
    margin-top:15px;
  }

  #content .date {
    margin-left:3px;
  }


  .highlight {
    margin: 10px 0;
  }

  .posts_listing {
    margin:0 0 50px;
  }

  .posts_listing li {
    margin:0 0 25px 15px;
  }

 
  #nav {
    text-align:center;
    position:static;
    margin-top:60px;
  }

  a {
    text-decoration: underline;
    text-decoration-color: #CCC;
    text-underline-offset: 5px;
  }

  a:hover {
    text-decoration-color: black;
  }

  
  #nav li {
    list-style-type:none;
    display:table-cell;
    padding: 0 20px;
  }

  #links {
    display: flex;
    justify-content: space-between;
    margin: 50px 0 0 0;
  }

  #links :nth-child(1) {
    margin-right:0.5em;;
  }

  #links :nth-child(2) {
    margin-left:0.5em;;
  }

  #not-found {
    text-align: center;
  }

  #not-found a {
    font-size: 3em;
    text-decoration: none;
    display: inline-block;
    padding-top: 225px;
    padding-bottom: 225px;
  }


#soft-info table td {
  padding: 2px 10px 2px 0px;
}

#license {
  font-style: italic;
  padding-top: 10px;
  padding-bottom: 10px;
  color:#403c3b;
}
#license a {
  color:#403c3b;
}
  @media (max-width: 750px) {
    body {
      padding-left:20px;
      padding-right:20px;
    }

    #nav h1 a {
      font-size:28px;
    }

    #nav li {
      font-size:13px;
      padding: 0 5px;
    }

    #content {
      margin-top:0;
      padding-top:50px;
      font-size:14px;
    }

    #content h1 {
      font-size:25px;
    }

    #content h2 {
      font-size:22px;
    }

    .posts_listing li div {
      font-size:12px;
    }
  }

  @media (max-width: 400px) {
    body {
      padding-left:20px;
      padding-right:20px;
    }

    #nav h1 a {
      font-size:22px;
    }

    #nav li {
      font-size:12px;
      padding: 0 5px;
    }

    #content {
      margin-top:0;
      padding-top:20px;
      font-size:12px;
    }

    #content h1 {
      font-size:20px;
    }

    #content h2 {
      font-size:18px;
    }

    .posts_listing li div{
      font-size:12px;
    }
  }

  @media (prefers-color-scheme: dark) {
    *, #nav h1 a {
      color: #FDFDFD;
    }

    body {
      background: #121212;
    }

    pre, code {
      background-color: #262626;
    }

    #sub-header, .date {
      color: #BABABA;
    }

    hr {
      background: #EBEBEB;
    }
  }
</style>


    </head>
<style>
#nav-logo-text {
    float: left;
    margin-top: 0px;
    margin-left: 0px;
    font-size: 2em;
    font-weight: 100;
}
#at {
  font-size:24px;
  padding:0px 2px;
}
</style>
<script>
window.root_dir = "lab-cn\/";
const regex = /^https?:\/\/[^/]+?\//;
root_dir = root_dir.replace(regex, "/");
$( function() {
  $("#content img").each(function(i) {
        var old_src = $(this).attr("src");
        var pattern = new RegExp("^" + root_dir);
        if(!pattern.test(old_src)) {
            pattern = /^http(s?):\/\//;
            if(!pattern.test(old_src)) {
                $(this).attr("src", root_dir + "/" + old_src);
            }
        }
    });
  $("#nav-logo-text").attr("href", root_dir);
  $("#no-found-a").attr("href", root_dir);

  var url = window.location.href;
  var pattern = /lab-cn/;
  var url2 = url.replace(pattern, "lab-en");
  $("#nav ul li:last-child a").attr("href", url2);
})
</script>
  <body>
    <section id=nav>
      <h1><a class="h-card" href="/"></a></h1>
      <a href="#" id="nav-logo-text">Gu<span style="padding:0px 2px">.</span>Z<span id="at">@</span>SUAT-SZ</a>
      <ul style="float:right">
        
          <li><a href="lab-cn/research/">研究</a></li>
        
          <li><a href="lab-cn/publication/">论文</a></li>
        
          <li><a href="lab-cn/software/">软件</a></li>
        
          <li><a href="lab-cn/post">博客</a></li>
        
          <li><a href="lab-cn/members/">成员</a></li>
        
          <li><a href="lab-cn/opening/">职位</a></li>
        
          <li><a href="lab-en/">EN</a></li>
        
      </ul>
      <div style="clear:both"></div>
    </section>


<section id=content>
  <h1 style="margin-bottom: 40px"> Visualize relations between genes and pathways </h1>

  
    <div id=sub-header>
      Zuguang Gu · November 2020 · <a href="#license">版权信息</a>
    </div>
  

  <div class="entry-content">
    <p>@<a href="https://github.com/venkan">venkan</a> provides <a href="https://github.com/jokergoo/circlize/issues/236">an interesting use
case</a> for <strong>circlize</strong>
which is to visualize the relations bewteen genes and pathways with
integrating genes&rsquo; genomic positions. An example plot looks like:</p>
<img src="https://i.stack.imgur.com/8mQv4.jpg" width="500" />
<p>In that plot, the &ldquo;pathway&rdquo; is in the same track as the ideograms. Thus, to
use <strong>circlize</strong> to implement it, we need to make &ldquo;pathway&rdquo; as a fake
chromosome and concatenate it to the normal chromosomes. In this blog post, I
will demonstrate how to implement it.</p>
<p>I first generate a random set of genes and their genomic positions. To make it
easier, I use single positions to represent genes (where what <code>df[, 3] = df[, 2]</code> does in the following code). I also assign pathway names for each gene.</p>
<pre><code class="language-r">set.seed(123)
library(circlize)
df = generateRandomBed(30)[ , 1:3]
df[, 3] = df[, 2]
df$gene = paste0(&quot;gene_&quot;, 1:nrow(df))
df$pathway = paste0(&quot;pathway_&quot;, sample(10, nrow(df), replace = TRUE))
head(df)
</code></pre>
<pre><code>##    chr     start       end   gene   pathway
## 1 chr1  12716970  12716970 gene_1 pathway_1
## 2 chr1 120710262 120710262 gene_2 pathway_5
## 3 chr1 161401295 161401295 gene_3 pathway_5
## 4 chr2  62435676  62435676 gene_4 pathway_8
## 5 chr2  66649737  66649737 gene_5 pathway_5
## 6 chr2 168175018 168175018 gene_6 pathway_7
</code></pre>
<p>Next I construct the concatenated genome by simply adding one more row to the
cytoband data frame. The newly added row corresponds to the fake-chromosome
&ldquo;pathway&rdquo; and the length of it corresponds to the length which will be finally
represented in the circular plot.</p>
<pre><code class="language-r">cytoband = read.cytoband()$df
cytoband = rbind(cytoband,
	data.frame(V1 = &quot;pathway&quot;, V2 = 1,  V3 = 2e8, V4 = &quot;&quot;, V5 = &quot;&quot;)
)
tail(cytoband)
</code></pre>
<pre><code>##           V1       V2        V3      V4     V5
## 858     chrY 15100000  19800000 q11.221 gpos50
## 859     chrY 19800000  22100000 q11.222   gneg
## 860     chrY 22100000  26200000 q11.223 gpos50
## 861     chrY 26200000  28800000  q11.23   gneg
## 862     chrY 28800000  59373566     q12   gvar
## 1100 pathway        1 200000000
</code></pre>
<p>Now &ldquo;pathway&rdquo; is a chromosome, and all the pathways (with name &ldquo;pathway_1&rdquo;,
&ldquo;pathway_2&rdquo;, &hellip;) are the &ldquo;genes&rdquo; in it. Next I calculate the &ldquo;positions&rdquo; of
these fake genes. I simply put the &ldquo;pathway genes&rdquo; evenly along the &ldquo;pathway&rdquo;
chromosome.</p>
<pre><code class="language-r">foo = round(seq(1, 2e8, length = 11))
pathway_mid = structure(as.integer((foo[1:10] + foo[2:11])/2), 
    names = paste0(&quot;pathway_&quot;, 1:10))

df$pathway_chr = &quot;pathway&quot;
df$pathway_start = pathway_mid[ df$pathway ]
df$pathway_end = pathway_mid[ df$pathway ]
head(df)
</code></pre>
<pre><code>##    chr     start       end   gene   pathway pathway_chr pathway_start
## 1 chr1  12716970  12716970 gene_1 pathway_1     pathway      10000001
## 2 chr1 120710262 120710262 gene_2 pathway_5     pathway      90000000
## 3 chr1 161401295 161401295 gene_3 pathway_5     pathway      90000000
## 4 chr2  62435676  62435676 gene_4 pathway_8     pathway     150000000
## 5 chr2  66649737  66649737 gene_5 pathway_5     pathway      90000000
## 6 chr2 168175018 168175018 gene_6 pathway_7     pathway     130000000
##   pathway_end
## 1    10000001
## 2    90000000
## 3    90000000
## 4   150000000
## 5    90000000
## 6   130000000
</code></pre>
<p>Everything is ready, and we can use the &ldquo;normal&rdquo; way to draw the genomic circular plot
with genomic links:</p>
<pre><code class="language-r">circos.initializeWithIdeogram(cytoband)
circos.genomicLink(df[, 1:3], df[, 6:8])
</code></pre>
<img src="/lab-cn/post/2020-11-27-visualize-relations-between-genes-and-pathways_files/figure-html/unnamed-chunk-5-1.png" width="768" />
<p>That is basically the idea. Next I customize the plot to make it look nicer, <em>e.g.</em>, to
adjust font size, set colors, add labels, &hellip;</p>
<pre><code class="language-r">circos.par(gap.after = c(rep(1, 23), 5, 5))
circos.genomicInitialize(cytoband, plotType = NULL)

# the labels track
label_df = rbind(df[, 1:4], setNames(df[, c(6:8, 5)], colnames(df)[1:4]))
label_df = unique(label_df)
circos.genomicLabels(label_df, labels.column = 4, side = &quot;outside&quot;, cex = 0.6)

# the chromosome names track
circos.track(track.index = get.current.track.index(), panel.fun = function(x, y) {
	circos.text(CELL_META$xcenter, CELL_META$ylim[1], CELL_META$sector.index,
		niceFacing = TRUE, adj = c(0.5, 0), cex = 0.8)
}, track.height = strheight(&quot;fj&quot;, cex = 0.8)*1.2, bg.border = NA, cell.padding = c(0, 0, 0, 0))

# ideogram track
circos.genomicIdeogram(cytoband)

# add points to the cell &quot;pathway&quot;
tb = table(df$pathway)
pathway_col = structure(1:10, names = paste0(&quot;pathway_&quot;, 1:10))
set.current.cell(sector.index = &quot;pathway&quot;, track.index = get.current.track.index())
circos.points(x = pathway_mid[names(tb)], y = CELL_META$ycenter, pch = 16,
	cex = tb/5, col = pathway_col[names(tb)])

# genomic links
circos.genomicLink(df[, 1:3], df[, 6:8], col = pathway_col[df[, 5]])
</code></pre>
<img src="/lab-cn/post/2020-11-27-visualize-relations-between-genes-and-pathways_files/figure-html/unnamed-chunk-7-1.png" width="768" />
<pre><code class="language-r">circos.clear()
</code></pre>

  </div>


  <p id="license">本文使用 <a href='https://creativecommons.org/licenses/by-nc-sa/4.0/deed.zh-hans'>CC BY-NC-SA 4.0</a> 协议发布。</p>


</section>


<style>
.footer p{
	color: #AAA;
	text-align: left;
	text-decoration: none;
	max-width:400px;
	border-top: 1px solid #AAA;
	padding-top:10px;
	font-size: 0.8em;
}
.footer p a {
	color: #AAA;
	text-align: left;
	text-decoration: none;
}
</style>

<footer class="footer">
<p>网站使用 <a href="https://gohugo.io/">Hugo</a>，<a href="https://bookdown.org/yihui/blogdown/">blogdown</a>，<a href="https://github.com/davidhampgonsalves/hugo-black-and-light-theme">Black & Light 主题</a>创建</p>
</footer>





<script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>



<script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
<script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/yaml.min.js"></script>
<script>hljs.configure({languages: []}); hljs.initHighlightingOnLoad();</script>




<script src="/js/math-code.js"></script>
<script async src="//cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML"></script>


</body>
</html>



